docs/docs/reference/other-new-features/open-classes.md

```diff
@@ -0,0 +1,79 @@
+}
+
+// File EncryptedWriter.scala
+package p
+
+class EncryptedWriter[T: Encryptable] extends Writer[T] {
+  override def send(x: T) = super.send(encrypt(x))
+}
+```
+An open class typically comes with some documentation that describes
+the internal calling patterns between methods of the class as well as hooks that can be overridden. We call this the _extension contract_ of the class. It is different from the _external contract_ between a class and its users.
+
+Classes that are not open can still be extended, but only if at least one of two alternative conditions is met:
+
+ - The extending class is in the same source file as the extended class. In this case, the extension is usually an internal implementation matter.
+
+ - The language feature `adhocExtensions` is enabled for the extending class. This is typically enabled by an import statement in the source file of the extension:
+   ```scala
+   import scala.language.adhocExtensions
+   ```
+   Alternatively, the feature can be enabled by the command line option `-language:adhocExtensions`.
+   If the feature is not enabled, the compiler will issue a "feature" warning. For instance, if the `open` modifier on class `Writer` is dropped, compiling `EncryptedWriter` would produce a warning:
+   ```
+   -- Feature Warning: EncryptedWriter.scala:6:14 ----
+     |class EncryptedWriter[T: Encryptable] extends Writer[T]
+     |                                              ^
+     |Unless class Writer is declared 'open', its extension in a separate file should be enabled
+     |by adding the import clause 'import scala.language.adhocExtensions'
+     |or by setting the compiler option -language:adhocExtensions.
+   ```
+
+### Motivation
+
+When writing a class, there are three possible expectations of extensibility:
+
+- The class is intended to allow extensions. This means one should expect
+a carefully worked out and documented extension contract for the class.
+
+- Extensions of the class are forbidden, for instance to make correctness or security guarantees.
+
+- There is no firm decision either way. The class is not _a priori_ intended for extensions, but if others find it useful to extend on an _ad-hoc_ basis, let them go ahead. However, they are on their own in this case. There is no documented extension contract, and future versions of the class might break the extensions (by rearranging internal call patterns, for instance).
+
+The three cases are clearly distinguished by using `open` for (1), `final` for (2) and no modifier for (3).
+
+It is good practice to avoid _ad-hoc_ extensions in a code base, since they tend to lead to fragile systems that are hard to evolve. But there
+are still some situations where these extensions are useful: for instance,
+to mock classes in tests, or to apply temporary patches that add features or fix bugs in library classes. That's why _ad-hoc_ extensions are permitted, but only if there is an explicit opt-in via a language feature import.
+
+### Details
```

Comment:
(I admit I have no idea what Lightbend's plan is for Scala and Modules, so forgive me if this is totally out of scope)

Would it make sense to automatically add a "module export" for anything that's considered `open`? I mean a [JSR 376](https://openjdk.java.net/projects/jigsaw/spec/sotms) `module { exports ... }` declaration.

Although perhaps this is an SBT concern and not something that the Scala compiler cares about....

Reply:
At present it is unclear to what extent JDK modules should affect the language (vs build tools).